internal/core/core.go

```diff
@@ -0,0 +1,272 @@
+	}
+	if _, ok := vendorCoreFileNamePattern[dutVendor]; !ok {
+		return nil, fmt.Errorf("add support for vendor %v in var vendorCoreFileNamePattern", dutVendor)
+	}
+	gClients, err := dut.DialGNOI(context.Background(), grpc.WithBlock())
+	if err != nil {
+		return nil, err
+	}
+	return &checker{
+		dut:        dut,
+		fileClient: gClients.File(),
+		initial:    coreFiles{},
+		delta:      coreFiles{},
+		startTime:  time.Now(),
+	}, nil
+}
+
+func (c *checker) check() (coreFiles, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	cores, err := c.checkCores()
+	if err != nil {
+		return nil, err
+	}
+	if len(c.initial) == 0 {
+		for k, v := range cores {
+			v.Existing = true
+			c.initial[k] = v
+		}
+		return c.initial, nil
+	}
+	for k, v := range cores {
+		if _, ok := c.initial[k]; !ok {
+			c.delta[k] = v
+		}
+	}
+	return c.delta, nil
+}
+
+type validatorImpl struct {
+	mu   sync.Mutex
+	duts map[string]*checker
+}
+
+func (v *validatorImpl) check() map[string]dutCoreFiles {
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+	dutCores := map[string]dutCoreFiles{}
+	for _, c := range v.duts {
+		wg.Add(1)
+		go func(c *checker) {
+			defer wg.Done()
+			cores, err := c.check()
+			status := "OK"
+			if err != nil {
+				status = fmt.Sprintf("DUT %q failed to check cores: %v", c.dut.Name(), err)
+				glog.Warning(status)
+			}
+			mu.Lock()
+			dutCores[c.dut.Name()] = dutCoreFiles{
+				DUT:    c.dut.Name(),
+				Files:  cores,
+				Status: status,
+			}
+			mu.Unlock()
+		}(c)
+	}
+	wg.Wait()
+	return dutCores
+}
+
+// start starts a core file watcher for the provided DUT.
+func (v *validatorImpl) start(duts map[string]binding.DUT) map[string]dutCoreFiles {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+	for k, dut := range duts {
+		glog.Infof("Registering core file checking for DUT %q", k)
+		c, err := newChecker(dut)
+		if err != nil {
+			glog.Warningf("Failed to register core file checking for DUT %q: %v", k, err)
+			continue
+		}
+		v.duts[k] = c
+	}
+	return v.check()
+}
+
+// Stop ends the validator and returns a list of all DUTs that
+// found core files.
+func (v *validatorImpl) stop() map[string]dutCoreFiles {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+	return v.check()
+}
+
+func registerBefore(e *eventlis.BeforeTestsEvent) error {
+	cores := validator.start(e.Reservation.DUTs)
+	ondatra.Report().AddSuiteProperty("validator.core", "enabled")
+	report := createReport(cores)
+	ondatra.Report().AddSuiteProperty("validator.core.initial", report)
+	return nil
+}
+
+const (
+	coreFmt = `
+Delta Core Files by DUT:{{range $key, $dut := .}} 
+DUT: {{$key}}{{ range $key, $cores := $dut.Files }}
+  {{ $key }}{{ end }}{{ end }}`
+)
+
+var coreTemplate = template.Must(template.New("errorMsg").Parse(coreFmt))
+
+func createReport(d map[string]dutCoreFiles) string {
+	b := new(bytes.Buffer)
+	if err := coreTemplate.Execute(b, d); err != nil {
+		fmt.Fprintf(b, "parse error on retrieving core files: %v", err)
```

Comment:
If `Execute` returns an error, b may have partial results. Do we want that partial string in the report in that case, or would it be better to return something like "<error>", or even just propagate the error?